below/ethtool/src/reader.rs

```diff
@@ -0,0 +1,232 @@
+                .map_err(|err| Error::ParseError(err.to_string()))
+        })
+        .collect::<Result<Vec<String>, Error>>()?;
+
+    Ok(names)
+}
+
+/// Parses the byte array returned by ioctl for ETHTOOL_GSTATS command.
+/// In case of error during parsing any feature,
+/// the function returns a `ParseError`.
+fn parse_values(
+    data: [u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+    length: usize,
+) -> Result<Vec<u64>, Error> {
+    let mut values = Vec::with_capacity(length);
+    let mut value_bytes = [0u8; 8];
+    for i in 0..length {
+        let offset = 8 * i;
+        match data.get(offset..offset + 8) {
+            Some(slice) => {
+                value_bytes.copy_from_slice(slice);
+                values.push(u64::from_ne_bytes(value_bytes));
+            }
+            None => {
+                return Err(Error::ParseError(format!(
+                    "parse value failed at offset={}",
+                    offset
+                )))
+            }
+        }
+    }
+
+    Ok(values)
+}
+
+/// A trait for reading stats using ethtool.
+///
+/// This trait allows mocking the ethtool calls for unit testing.
+pub trait EthtoolReadable {
+    fn new(if_name: &str) -> Result<Self, Error>
+    where
+        Self: Sized;
+    fn stats(&self) -> Result<Vec<(String, u64)>, Error>;
+}
+
+pub struct Ethtool {
+    sock_fd: OwnedFd,
+    if_name: [u8; IFNAME_MAX_SIZE],
+}
+
+impl Ethtool {
+    /// Get the number of stats using ETHTOOL_GSSET_INFO command
+    fn gsset_info(&self) -> Result<usize, Errno> {
+        let mut sset_info = StringSetInfo {
+            cmd: ETHTOOL_GSSET_INFO,
+            reserved: 1,
+            mask: 1 << ETH_SS_STATS,
+            data: 0,
+        };
+
+        match _ioctl(
+            &self.sock_fd,
+            self.if_name,
+            &mut sset_info as *mut StringSetInfo as usize,
+        ) {
+            Ok(_) => Ok(sset_info.data),
+            Err(errno) => Err(errno),
+        }
+    }
+
+    /// Get the feature names using ETHTOOL_GSTRINGS command
+    fn gstrings(&self, length: usize) -> Result<Vec<String>, Error> {
+        let mut gstrings = GStrings {
+            cmd: ETHTOOL_GSTRINGS,
+            string_set: ETH_SS_STATS,
+            len: length as u32,
+            data: [0u8; MAX_GSTRINGS * ETH_GSTRINGS_LEN],
+        };
+
+        match _ioctl(
+            &self.sock_fd,
+            self.if_name,
+            &mut gstrings as *mut GStrings as usize,
+        ) {
+            Ok(_) => return parse_names(gstrings.data, length),
+            Err(errno) => Err(Error::GStringsReadError(errno)),
+        }
+    }
+
+    /// Get the statistics for the features using ETHTOOL_GSTATS command
+    fn gstats(&self, features: &[String]) -> Result<Vec<u64>, Error> {
+        let length = features.len();
+        let mut gstats = GStats {
+            cmd: ETHTOOL_GSTATS,
+            len: features.len() as u32,
+            data: [0u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+        };
```

Comment:
I have similar concerns here as with gstrings, this doesn't match my understanding of the API